File: packages/crawler-py/shroodler/baseline.py
from __future__ import annotations

from urllib.parse import urlparse

from shroodler.suppress import filter_findings, path_of


def _page_path(url: str) -> str:
    return path_of(url)


def _origin_host(url: str) -> str:
    p = urlparse(url)
    if p.scheme and p.netloc:
        return f"{p.scheme}://{p.netloc}"
    return url


def document_to_baseline(
    doc: dict,
    *,
    name: str | None = None,
    suppressions: list[dict] | None = None,
) -> dict:
    pages = doc.get("pages") or []
    findings = filter_findings(doc.get("findings") or [], suppressions or [])
    expected_pages = sorted({_page_path(p.get("url", "")) for p in pages})
    expected_forms: dict[str, list[str]] = {}
    for page in pages:
        path = _page_path(page.get("url", ""))
        names: set[str] = set()
        for form in page.get("forms") or []:
            for field in form.get("fields") or []:
                n = field.get("name")
                if n:
                    names.add(n)
        if names:
            existing = set(expected_forms.get(path, []))
            expected_forms[path] = sorted(existing | names)
    expected_findings = sorted(
        ({"id": f.get("id", ""), "url": _page_path(f.get("url", ""))} for f in findings),
        key=lambda row: (row["id"], row["url"]),
    )
    target = doc.get("target") or ""
    return {
        "target_app": name or target or "local-app",
        "target": target,
        "expected_pages": expected_pages,
        "expected_forms": dict(sorted(expected_forms.items())),
        "expected_findings": expected_findings,
        "expected_not_found": [],
    }


def origin_of(url: str) -> str:
    return _origin_host(url)

File: packages/desktop-app/src/lib/baseline.js
export function pathOf(url) {
  try {
    return new URL(url).pathname || "/";
  } catch {
    return "/";
  }
}

export function documentToBaseline(doc, { name, suppressions = [] } = {}) {
  const pages = doc.pages || [];
  const findings = (doc.findings || []).filter((f) => !isSuppressed(f, suppressions));
  const expectedPages = [...new Set(pages.map((p) => pathOf(p.url)))].sort();
  const forms = {};
  for (const page of pages) {
    const path = pathOf(page.url);
    const names = new Set(forms[path] || []);
    for (const form of page.forms || []) {
      for (const field of form.fields || []) {
        if (field.name) names.add(field.name);
      }
    }
    if (names.size) forms[path] = [...names].sort();
  }
  const expectedFindings = findings
    .map((f) => ({ id: f.id || "", url: pathOf(f.url) }))
    .sort((a, b) => a.id.localeCompare(b.id) || a.url.localeCompare(b.url));
  const target = doc.target || "";
  return {
    target_app: name || target || "local-app",
    target,
    expected_pages: expectedPages,
    expected_forms: Object.fromEntries(Object.entries(forms).sort(([a], [b]) => a.localeCompare(b))),
    expected_findings: expectedFindings,
    expected_not_found: [],
  };
}

export function globMatch(pattern, value) {
  if (!pattern || pattern === "*") return true;
  const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*").replace(/\?/g, ".");
  return new RegExp(`^${escaped}$`).test(value);
}

export function parseSuppressions(raw) {
  const text = String(raw || "").trim();
  if (!text) return [];
  if (text.startsWith("{") || text.startsWith("[")) {
    try {
      const data = JSON.parse(text);
      const list = Array.isArray(data) ? data : data.suppressions || [];
      return list.map((row) => ({
        id: row.id || "*",
        url: row.url || "*",
        reason: row.reason || "",
      }));
    } catch {
      return [];
    }
  }
  const rows = [];
  const blocks = text.split(/\n\s*-\s+/);
  for (const chunk of blocks) {
    const id = (chunk.match(/(?:^|\n)\s*id:\s*(\S+)/) || [])[1];
    const url = (chunk.match(/(?:^|\n)\s*url:\s*(\S+)/) || [])[1];
    const reasonM = chunk.match(/(?:^|\n)\s*reason:\s*(.*)$/m);
    if (!id && !url) continue;
    rows.push({
      id: (id || "*").replace(/^['"]|['"]$/g, ""),
      url: (url || "*").replace(/^['"]|['"]$/g, ""),
      reason: reasonM ? reasonM[1].trim().replace(/^["']|["']$/g, "") : "",
    });
  }
  return rows;
}

export function isSuppressed(finding, rules) {
  const path = pathOf(finding.url);
  for (const rule of rules || []) {
    if (rule.id !== "*" && rule.id !== finding.id) continue;
    if (globMatch(rule.url, path) || globMatch(rule.url, finding.url || "")) return rule;
  }
  return null;
}

export function suppressionsYaml(rules) {
  const lines = ["suppressions:"];
  for (const r of rules || []) {
    lines.push(`  - id: ${r.id || "*"}`);
    lines.push(`    url: ${r.url || "*"}`);
    if (r.reason) lines.push(`    reason: ${JSON.stringify(r.reason)}`);
  }
  if (!rules?.length) lines.push("  []");
  return lines.join("\n") + "\n";
}
